controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationFile.java

```diff
@@ -367,7 +367,16 @@ private File determineMainFile(final String rawName, final String name) {
             mainName = stripPrefixSuffix(name);
         }
         if (mainName != null) {
-            return new File(configurationDir, new File(mainName).getName());
+            try {
+                final File mainFile = new File(configurationDir, name != null ? name : mainName);
+                if (mainFile.getCanonicalPath().startsWith(configurationDir.getCanonicalPath())) {
+                    return new File(configurationDir, new File(mainName).getName());
+                } else if (interactionPolicy.isReadOnly()) {
+                    return mainFile;
+                }
+            } catch (IOException ioe) {
+                throw ControllerLogger.ROOT_LOGGER.canonicalMainFileNotFound(ioe, mainFile);
+            }
```

Comment:
@yersan

> […] this method is not going to return the file from where read the boot operations.

Right, but when using READ_ONLY, on line 362 we do `return absoluteFile` so I wanted to return "relativeFile" in this situation to keep it consistent

So we're either returning 
 * `new File(configurationDir, new File(mainName).getName())` as before, or
 * `new File(configurationDir, name)` only in the case where the file is not in the config folder and `interactionPolicy.isReadOnly()`

Reply:
Hi @michpetrov, yes, but that's for a different case, it returns the absolute path if directoryFile doesn't exist L356, otherwise  the old code will reach the L370, returning `new File(configurationDir, new File(mainName).getName())`.

I'm I'm not wrong, with your code we are introducing a side change, if directoryFile.exists(), the path is outside of the config dir and interactionPolicy.isReadOnly(), now the method could return:
* `new File(configurationDir, name)`
* `new File(configurationDir, mainName)`

Depending whether `name` is null or not.

Is not this a change beyond to what the issue is trying to fix?

I don't have all the expertise regarding this subject, if you think that is fine, the test are passing and it should be fine, but I just only trying to highlight this new behaviour in case there is any kind of doubts.

Reply:
The only change is when a relative path ("../tmp/custom.xml") is used with READ_ONLY
 * current ⇒ "CONFIG_DIR/custom.xml" (this file does not exist or it is a different file)
 * new ⇒ "CONFIG_DIR/../tmp/custom.xml" (this file exists)

L375 will always return `new File(configurationDir, name)`; if `name` is null then `mainName` is assigned `rawName` (L340) and that will be inside the config folder so we return at L373

Reply:
> The only change is when a relative path ("../tmp/custom.xml") is used with READ_ONLY
> 
> * current ⇒ "CONFIG_DIR/custom.xml" (this file does not exist or it is a different file)
> * new ⇒ "CONFIG_DIR/../tmp/custom.xml" (this file exists)
> 
Exactly, so this change is introducing a new behavior in this method. As I suggested before, I think it can be avoided by using https://github.com/wildfly/wildfly-core/pull/4947#discussion_r819726859. 

That was just only a suggestion, I don't have arguments in favor of your changes but I also don't have arguments against them. The tests pass, this new behavior is intentional and I assume you have tested this case manually, so I'm fine with it. Thanks for your patience in discussing it @michpetrov.